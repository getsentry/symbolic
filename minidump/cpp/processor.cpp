#include "google_breakpad/processor/basic_source_line_resolver.h"
#include "google_breakpad/processor/minidump.h"
#include "google_breakpad/processor/minidump_processor.h"
#include "google_breakpad/processor/process_state.h"

#include "cpp/data_definitions.h"
#include "cpp/memstream.h"
#include "cpp/mmap_symbol_supplier.h"
#include "cpp/processor.h"

using google_breakpad::BasicSourceLineResolver;
using google_breakpad::Minidump;
using google_breakpad::MinidumpProcessor;
using google_breakpad::ProcessState;

process_state_t *process_minidump(const char *buffer,
                                  size_t buffer_size,
                                  symbol_entry_t *symbols,
                                  size_t symbol_count,
                                  int *result_out) {
    if (buffer == nullptr) {
        *result_out = google_breakpad::PROCESS_ERROR_MINIDUMP_NOT_FOUND;
        return nullptr;
    }

    ProcessState *state = new ProcessState();
    if (state == nullptr) {
        *result_out = -1;  // Memory allocation issue
        return nullptr;
    }

    BasicSourceLineResolver resolver;
    MmapSymbolSupplier supplier(symbol_count, symbols);
    MinidumpProcessor processor(&supplier, &resolver);

    imemstream in(buffer, buffer_size);
    Minidump minidump(in);
    if (!minidump.Read()) {
        *result_out = google_breakpad::PROCESS_ERROR_MINIDUMP_NOT_FOUND;
        return nullptr;
    }

    *result_out = processor.Process(&minidump, state);
    if (*result_out != google_breakpad::PROCESS_OK) {
        delete state;
        return nullptr;
    }

    return process_state_t::cast(state);
}
