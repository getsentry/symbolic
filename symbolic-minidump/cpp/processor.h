#ifndef SENTRY_PROCESSOR_H
#define SENTRY_PROCESSOR_H

#include <cstddef>
#include "cpp/data_structures.h"

#ifdef __cplusplus
extern "C" {
#endif

/// Data transfer object for symbols in memory
struct symbol_entry_t {
    /// The debug identifier of the code module these symbols are for
    const char *debug_identifier;

    /// Size of the buffer inside symbol_data
    const size_t symbol_size;

    /// Raw data of the symbol file passed to the symbolizer
    const char *symbol_data;
};

/// Reads a minidump from a memory buffer and processes it. Returns an owning
/// pointer to a process_state_t struct that contains loaded code modules and
/// call stacks of all threads of the process during the crash.
///
/// Processing the minidump can fail if the buffer is corrupted or does not
/// exit. The function will return NULL and an error code in result_out.
///
/// Release memory of the process state with process_state_delete.
process_state_t *process_minidump(const char *buffer,
                                  size_t buffer_size,
                                  symbol_entry_t *symbols,
                                  size_t symbol_count,
                                  int *result_out);

#ifdef __cplusplus
}
#endif

#endif
