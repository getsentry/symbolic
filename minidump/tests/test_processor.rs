use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_minidump::processor::ProcessState;

#[test]
fn process_minidump_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_linux", &state);
    Ok(())
}

#[test]
fn process_minidump_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_macos", &state);
    Ok(())
}

#[test]
fn process_minidump_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_windows", &state);
    Ok(())
}

#[test]
fn get_referenced_modules_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("referenced_modules_linux", &state.referenced_modules());
    Ok(())
}

#[test]
fn get_referenced_modules_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("referenced_modules_macos", &state.referenced_modules());
    Ok(())
}

#[test]
fn get_referenced_modules_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!(
        "referenced_modules_windows",
        &state.referenced_modules()
    );
    Ok(())
}
