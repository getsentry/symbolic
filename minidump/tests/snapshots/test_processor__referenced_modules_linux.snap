---
created: "2019-02-20T15:42:12.012631Z"
creator: insta@0.6.3
source: minidump/tests/test_processor.rs
expression: "&state.referenced_modules()"
---
{
    CodeModule {
        id: Some(
            CodeModuleId {
                inner: DebugId {
                    uuid: "451a38b5-0679-79d2-0738-22a5ceb24c4b",
                    appendix: 0
                }
            }
        ),
        base_address: 139986944249856,
        size: 1835008,
        code_file: "/lib/x86_64-linux-gnu/libc-2.23.so",
        code_identifier: "b5381a457906d279073822a5ceb24c4bfef94ddb",
        debug_file: "/lib/x86_64-linux-gnu/libc-2.23.so",
        debug_identifier: "451A38B5067979D2073822A5CEB24C4B0"
    },
    CodeModule {
        id: Some(
            CodeModuleId {
                inner: DebugId {
                    uuid: "59627b5d-2255-a375-c17b-d4c3fd05f5a6",
                    appendix: 0
                }
            }
        ),
        base_address: 139986956304384,
        size: 155648,
        code_file: "/lib/x86_64-linux-gnu/ld-2.23.so",
        code_identifier: "5d7b6259552275a3c17bd4c3fd05f5a6bf40caa5",
        debug_file: "/lib/x86_64-linux-gnu/ld-2.23.so",
        debug_identifier: "59627B5D2255A375C17BD4C3FD05F5A60"
    },
    CodeModule {
        id: Some(
            CodeModuleId {
                inner: DebugId {
                    uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                    appendix: 0
                }
            }
        ),
        base_address: 4194304,
        size: 106496,
        code_file: "/work/linux/build/crash",
        code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
        debug_file: "/work/linux/build/crash",
        debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
    }
}
